### ISSUE TITLE
Status Code:204 No Content Remote

### ISSUE THREAD

**Issue by steida** [facebook/relay#2333]
I don't know why my server API returns 204 for now, but if I add two items with the same text, Relay will generate the same key.

![screen shot 2018-02-21 at 04 46 24](https://user-images.githubusercontent.com/66249/36462764-9b00bcc8-16c5-11e8-8bb1-1665033c57c4.png)

**Comment by steida** [comment]
Maybe explicit clientMutationId with random ID prevented that. Idk.

**Comment by steida** [comment]
Jeez, wtf is this shit? Type fetch and json???

<img width="915" alt="screen shot 2018-02-21 at 05 26 42" src="https://user-images.githubusercontent.com/66249/36463084-d96c8be8-16c7-11e8-9809-a76fd66e98a1.png">

**Comment by alloy** [comment]
Isn’t the 204 possibly a response to a CORS preflight request?

As for the `key`, aren’t you in charge of rendering the component and in the possibility to render it with unique keys?

**Comment by alloy** [comment]
So is the `id` of your `node` the same but otherwise the `node` data is different?

**Comment by steida** [comment]
No. The error was caused by missing clientMutatationId. What is weird that server does not have to use it at all, but client must pass it with random number.

**Comment by alloy** [comment]
I’m not sure why you closed this and how the `clientMutationId` relates to your `node`’s `id` field. Does your server not populate the `id` field with a globally unique value?

**Comment by steida** [comment]
Until I add Relay input type shape with clientMutationId, Relay for some reason renders just created items with the same name prop with the same id, so React screams about a duplicated key.

https://github.com/este/este/blob/710f3fa0171ff48751e94ef4fe81a4ac8e4c7cb4/components/CreateWeb.js#L44

So, Relay somehow internally requires clientMutationId, othetwise, the rendering fails with the aforementioned error.

**Comment by steida** [comment]
@jstejada Is somewhere described when `clientMutationId` is required? Also, does Relay any magic with input type shape? Can you please point me to related Relay source code files? Thank you.

**Comment by steida** [comment]
clientMutationId is killing me. No docs, no dev warning, just weird React rendering bug...

**Comment by hisapy** [comment]
`clientMutationId` is not required in Relay modern. I removed all of them after upgrading from Relay Classic, also removing it from the server - Actually I submitted a pull request to the lib (Absinthe/Elixir) we use to generate the GraphQL schema to avoid generating the `clientMutationId` field given a certain _config_.  

That said, I have a couple debugging ideas/questions:

1. Did you run the component query/mutation outside Relay, perhaps GraphiQL ?
2. Did you try isolating the component testing it with static/mock data that supposed to have the correct shape? 

I use `relay-mock-network-layer` to write storyshots to _visually_ test components without depending on data from the server. It's super useful and much more fun than writing tests _manually_

**Comment by steida** [comment]
It is required. That's why I crated this issue. Without a clientMutationId, when I add two items with the same text, Relay fails. https://github.com/este/este/blob/Prisma/components/CreateWeb.js#L44

There is a bug in Relay.

As for 1. Sure I did. And it works, but RELAY FAILS OMG GRAPHIQL IS NOT RELATED TO RELAY, OMG.

As for 2. OMG I USE CONSOLE.LOG OMG I ADDED SCREENSHOT WHY ARE YOU ASKING?

Omg. Thank you.

**Comment by steida** [comment]
Let me rephrase it: clientMutationId is must otherwise Relay does not work. Thank you.

**Comment by steida** [comment]
It's a long-standing Relay bug. I bitched about it several times. Unless I add foking clientMutationId, client Relay update fail when some object with the same text (shape) is added. Omg.

**Comment by steida** [comment]
OK. I don't want to be rude. Yes, I am, I know. I will extend my developer death march for next few weeks reading the Relay source code, investigate foking clientMutationId curse. OK. This is OSS.

**Comment by hisapy** [comment]
No problem. FWIW, 

1. I can't reproduce this bug in my current code with **Relay Modern**. 
2. I know what's failing according to the issue. I was just throwing debugging ideas like isolating parts of the system, i.e.: I don't know if you get repeated ids when executing the mutations outside of Relay, so we can discard the server as a possible cause of the problem.

Good luck

**Comment by taion** [comment]
Yeah, this is the same issue as https://github.com/relayjs/relay-examples/issues/30, fixed there in https://github.com/relayjs/relay-examples/commit/ed955945ca5158ac34ed8ce703ae0cc7cb3e38e7#diff-4b42725e6eb503f56da8e3f637e9a4e3.

@wincent's response in that issue on https://github.com/relayjs/relay-examples/issues/30#issuecomment-306245016 specifically makes sense to me.

This is quite a rough edge, though, but it's a tricky trade-off. For now, maybe we should add an explicit caveat to https://facebook.github.io/relay/docs/en/mutations.html#commitmutation to note the importance of specifying a `clientMutationId`?

**Comment by steida** [comment]
I would rather see dev warning for Relay or remove it entirely. @kassens

**Comment by kassens** [comment]
Closing this in favor of the more specific clientMutationId issue.
